nx-X11/programs/Xserver/hw/nxagent/Screen.c

```diff
@@ -3674,24 +3674,292 @@ Bool intersect_bb(int ax1, int ay1, unsigned int aw, unsigned int ah,
+	     * in the bounding box x-wise.
+	     *
+	     * (Same calculation as above...)
+	     * right session window border    minus: left border of this CRTC
+	     *           |                       |
+	     *           v                       v
+	     */
+	    width = (ax1 + aw)                - bx1;
+	    #ifdef DEBUG
+	    fprintf(stderr, "intersect_bb: width = (ax1 + aw) - bx1 = [%d]\n", width);
+	    #endif
+	}
     }
 
-    if (ay1 < bby1 || ay1 + ah > bby2) {
-        #ifdef DEBUG
-        fprintf(stderr, "intersect: box has parts outside bounding box - height stays unchanged [%d]\n", ah);
-        #endif
-	*h = ah;
+    /* check if outside-right of bounding box */
+    else if (ax1 + aw > bbx2) {
+
+	#ifdef DEBUG
+	fprintf(stderr, "intersect_bb: box is outside-right of the bounding box - width gets adapted (see below)\n");
+	#endif
+
+	if (bx1 + bw < bbx2 && bx1 < ax1 && ax1 < bx1 + bw) {
+	    /* The session window width starts in this CRTC and
+	     * spreads over into the next CRTC to the right and
+	     * and beyond into the right side of the bounding
+	     * box.
+	     *
+	     * Left of this CRTC is assumed to be another CRTC.
+	     *
+	     * right border of this CRTC    minus: left border of session window
+	     *           |                       |
+	     *           v                       v
+	     */
+	    width = (bx1 + bw)                - ax1;
+	    #ifdef DEBUG
+	    fprintf(stderr, "intersect_bb: width = (bx1 + bw) - ax1 = [%d]\n", width);
+	    #endif
+	} else if (bx1 + bw < bbx2 && ax1 < bx1) {
+	    /* The session window width spreads across this
+	     * whole CRTC coming from the left into the one
+	     * on the right.
+	     */
+	    width = bw;
+	    #ifdef DEBUG
+	    fprintf(stderr, "intersect_bb: width = bw = [%d]\n", width);
+	    #endif
+	} else if (bx1 + bw == bbx2 && ax1 < bx1) {
+	    /* The session window width spreads across the whole
+	     * CRTC and has some rest-width outside-right of the
+	     * bounding box.
+	     *
+	     * right border of session window  minus: left border of CRTC
+	     *           |                       |
+	     *           v                       v
+	     */
+	    width = (ax1 + aw)                   - bx1;
+	    #ifdef DEBUG
+	    fprintf(stderr, "intersect_bb: width = (ax1 + aw) - bx1 = [%d]\n", width);
+	    #endif
+	}
     }
-  }
 
+    if ((ax1 < bbx1 || ax1 + aw > bbx2) && width == -1) {
+      #ifdef DEBUG
+      fprintf(stderr, "intersect_bb: width stays unchanged [%d]\n", aw);
+      #endif
+
+      /* Catch all other beyond-bounding-box cases.
+       *
+       * This mainly is: session window is x-wise on one CRTC only.
+       */
+      *w = aw;
+
+    } else if (ax1 < bbx1 || ax1 + aw > bbx2) {
+      #ifdef DEBUG
+      fprintf(stderr, "intersect_bb: bounding box adapted width gets set to [%d]\n", width);
+      #endif
+      *w = abs(width);
+    }
+
+    /*
+     * ###### The Y-Coordinate ######
```

Comment:
I have not checked all the details yet but from the quick glance I took this _should_ be basically the identical code to the X coordinate. So factor that out to a separate function.